cmd/gen-key/rsa.go

```diff
@@ -0,0 +1,132 @@
+package main
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"log"
+	"math/big"
+
+	"github.com/miekg/pkcs11"
+)
+
+func rsaArgs(label string, modLen, exp uint) ([]*pkcs11.Mechanism, []*pkcs11.Attribute, []*pkcs11.Attribute) {
+	expSlice := make([]byte, 8)
+	binary.BigEndian.PutUint64(expSlice, uint64(exp))
+	log.Printf("\tEncoded public exponent (%d) as: %0X\n", exp, expSlice)
+	return []*pkcs11.Mechanism{
+			pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS_KEY_PAIR_GEN, nil),
+		},
+		[]*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+			pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+			pkcs11.NewAttribute(pkcs11.CKA_VERIFY, true),
+			pkcs11.NewAttribute(pkcs11.CKA_MODULUS_BITS, modLen),
+			pkcs11.NewAttribute(pkcs11.CKA_PUBLIC_EXPONENT, expSlice),
+		}, []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+			pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+			// Prevent attributes being retrieved
+			pkcs11.NewAttribute(pkcs11.CKA_SENSITIVE, true),
+			// Prevent the key being extracted from the device
+			pkcs11.NewAttribute(pkcs11.CKA_EXTRACTABLE, false),
+			// Allow the key to sign data
+			pkcs11.NewAttribute(pkcs11.CKA_SIGN, true),
+		}
+}
+
+func rsaPub(ctx Ctx, session pkcs11.SessionHandle, object pkcs11.ObjectHandle, modLen, exp uint) (*rsa.PublicKey, error) {
+	// Retrieve the public exponent and modulus for the generated public key
+	attrs, err := ctx.GetAttributeValue(session, object, []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_PUBLIC_EXPONENT, nil),
+		pkcs11.NewAttribute(pkcs11.CKA_MODULUS, nil),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to retrieve key attributes: %s", err)
+	}
+
+	// Attempt to build the public key from the retrieved attributes
+	pubKey := &rsa.PublicKey{}
+	gotMod, gotExp := false, false
+	for _, a := range attrs {
+		switch a.Type {
+		case pkcs11.CKA_PUBLIC_EXPONENT:
+			pubKey.E = int(big.NewInt(0).SetBytes(a.Value).Int64())
```

Comment:
Above we use `binary.BigEndian` to encode the exponent; here we use `SetBytes` to decode it. It seems like it would e better to use just one package or the other.

Reply:
Hrm, my memory is that something was failing when I used an un-padded slice (like that returned by `big.Int.Bytes()`) but when I just tested it it worked fine on the hardware and SoftHSMv2 so ¯\_(ツ)_/¯ (It's possible that this failed with an earlier version of SoftHSMv2 I guess, I'm currently using master).

Probably it makes sense to just uniformly use `big` here.